src/test/FuzzerImpl.cpp

```diff
@@ -548,6 +595,85 @@ TransactionFuzzer::initialize()
     resetTxInternalState(*mApp);
     LedgerTxn ltxOuter(mApp->getLedgerTxnRoot());
 
+    std::array<FuzzUtils::AccountParameters,
+               FuzzUtils::NUMBER_OF_PREGENERATED_ACCOUNTS>
+        accountParameters;
+
+    // We don't issue assets from account 0.
+    accountParameters[0].mNumAssetsToDistribute = 0;
+
+    std::vector<FuzzUtils::OfferParameters> offerParameters;
+
+    {
+        // In order for certain operation combinations to be valid, we need an
+        // initial order book with various offers. The order book will consist
+        // of identical setups for the asset pairs:
+        //      XLM - A
+        //      A   - B
+        //      B   - C
+        //      C   - D
+        // For any asset A and asset B, the generic order book setup will be as
+        // follows:
+        //
+        // +------------+-----+------+--------+------------------------------+
+        // |  Account   | Bid | Sell | Amount | Price (in terms of Sell/Bid) |
+        // +------------+-----+------+--------+------------------------------+
+        // | 0          | A   | B    |      1 | 3/2                          |
+        // | 1 (issuer) | A   | B    |      5 | 3/2                          |
+        // | 2          | A   | B    |     10 | 1/1                          |
+        // | 3 (issuer) | B   | A    |      1 | 10/9                         |
+        // | 4          | B   | A    |      5 | 10/9                         |
+        // | 0          | B   | A    |     10 | 22/7                         |
+        // +------------+-----+------+--------+------------------------------+
+        //
+        // This gives us a good mixture of buy and sell, issuers with offers,
+        // and an account with a bid and a sell. It also gives us an initial
+        // order book that is in a legal state.
+
+        auto addAssetPairParameters = [&accountParameters,
+                                       &offerParameters](Asset A, Asset B,
+                                                         std::vector<int> pks) {
+            auto addOfferParameters = [&accountParameters, &offerParameters](
+                                          Asset bid, Asset sell, int pk,
+                                          Price price, int64 amount) {
+                offerParameters.emplace_back(
+                    FuzzUtils::OfferParameters(bid, sell, pk, price, amount));
+                ++accountParameters[pk].mNumOffers;
+                accountParameters[pk].mOfferLiabilities +=
```

Comment:
> You shouldn't be computing this directly. The actual underlying computations are complicated so this may not work in general. Use getOfferSellingLiabilities or getOfferBuyingLiabilities.

Thanks -- I've pushed a commit to use `getOfferBuyingLiabilities()`.

> Also this calculation doesn't make sense to me. You aren't distinguishing between liabilities for different assets.

Currently the only thing that I'm using `AccountParameters::mOfferLiabilities` for is to help compute the total balance that we're going to give an account when it's created during fuzzer initialization, which I thought would only depend upon the total liabilities for all assets.  Does it already need to be per-asset for some reason I haven't thought of?  (I can imagine that I might need to make it that way in the future to support more specifically-targeted setup parameters -- for example, perhaps, to try to guide the fuzzer towards test cases that would try to exceed trustline limits without exceeding account balances?)

Reply:
This is a fundamental misunderstanding of liabilities. I strongly encourage you to read [CAP-0003](https://github.com/stellar/stellar-protocol/blob/master/core/cap-0003.md) to get a more thorough understanding. Liabilities are always tied to the asset. You should think of them as having units equal to the asset, so adding them for different assets can never make sense. Specifically, the account balance should only be influenced by the native selling liabilities (eg. the total amount of native asset that would be sold if all offers selling native executed simultaneously).

Reply:
Okay, now that you say this I realize I'm adding meters to electron-volts.

> Specifically, the account balance should only be influenced by the native selling liabilities (eg. the total amount of native asset that would be sold if all offers selling native executed simultaneously).

Done.